infrasec/tutorials/atlantis.md

```diff
@@ -0,0 +1,418 @@
+EC2 > load balancer > `atlantis-exp`
+
+<img src="https://github.com/trussworks/Engineering-Playbook/blob/3efe6ea02ed010f3db2c07921c5c8acc60406b84/infrasec/tutorials/images/atlantis_alb2.png" width="450">
+
+We can click on our certificate and manually point to the load balancer.
+
+### VPC Troubleshooting
+
+Anecdotally, the variables for `private_subnets` and `public_subnets` did not succesfully connect to the VPC when plugging in the IP values directly. However, `private_subnet_ids` and `public_subnet_ids` work fine.
+
+There are no open issues in the module for this discrepancy, so this mystery may be confined to the project (or author). This information is included here simply for posterity.
+
+### ALB Troubleshooting
+
+We'll see [two ALB listeners](https://github.com/terraform-aws-modules/terraform-aws-atlantis/blob/124e266a8c5746b948114820c07339b5b221917d/main.tf#L212-L235) in the plan related to the redirect on the ALB. Here's a sample plan output:
+
+```hcl
+  # module.atlantis.module.alb.aws_lb_listener.frontend_http_tcp[0] will be created
+  ...
+      + port              = 80
+      + protocol          = "HTTP"
+      + ssl_policy        = (known after apply)
+
+  ...{{
+
+          + redirect {
+              + host        = "#{host}"
+              + path        = "/#{path}"
+              + port        = "443"
+              + protocol    = "HTTPS"
+              + query       = "#{query}"
+              + status_code = "HTTP_301"
+            }
+        }
+    }
+
+  # module.atlantis.module.alb.aws_lb_listener.frontend_https[0] will be created
+  + resource "aws_lb_listener" "frontend_https" {
+  ...
+      + port              = 443
+      + protocol          = "HTTPS"
+      + ssl_policy        = "ELBSecurityPolicy-2016-08"
+
+      + default_action {
+          + order            = (known after apply)
+          + target_group_arn = (known after apply)
+          + type             = "forward"
+        }
+    }
+```
+
+Check out these ports because this is what's happening:
+
+<img src="https://github.com/trussworks/Engineering-Playbook/blob/3efe6ea02ed010f3db2c07921c5c8acc60406b84/infrasec/tutorials/images/atlantis_alb1.png" width="450">
+
+The ALB is being created with these two listeners (one [https](https://github.com/terraform-aws-modules/terraform-aws-atlantis/blob/124e266a8c5746b948114820c07339b5b221917d/main.tf#L212-L222) & one [http](https://github.com/terraform-aws-modules/terraform-aws-atlantis/blob/124e266a8c5746b948114820c07339b5b221917d/main.tf#L224-L235)). The http port (on `80`) serves to redirect to https/`443` and force use of our ACM certificate, setting up SSL termination on the load balancer. This keeps us from having to jump through the hoops of setting up docker and the client with certificates and dealing with SSL termination the TCP way (which is also how we would have to terminate the certificate with NLBs).
+
```

Comment:
...what is this about troubleshooting? Did you have the target group not link to the listener when you were setting this up? This sounds...very strange.

Reply:
It did not and it was very weird. Barry had to pair with me to help me get me past this.

Reply:
That is very weird and doesn't feel like something wrong with this module...I suspect something else got out of sync. It's more of a general ALB troubleshooting issue anyway, which feels not appropriate for this particular tutorial.

Reply:
> doesn't feel like

Really? That's subjective. If something got out of sync, then it's related to usage of this module and it's appropriate.